### PR TITLE
[WIP] EIP165 interfaces for token standard abstraction

### PR DESCRIPTION
**Pull request by Recmo**
## Description

Add EIP165 interfaces and supporting logic.

More to follow.

## Motivation and Context

EIP165 introduces a standard for contract introspection. This allows a contract to deduce what kind of token standard (ERC20, ERC223, ERC821, etc) a given address supports (assuming the contract supports ERC165).

## How Has This Been Tested?

TBD

## Types of changes

* [x] Change requires a change to the documentation.
* [ ] Added tests to cover my changes.

### CODE REVIEW

File: packages/contracts/contracts/EIP165Cache.sol [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+// https://github.com/jbaylina/EIP165Cache/blob/master/contracts/EIP165Cache.sol
+//
+// Changes:
+// * Optimizations
+// * Allow cache refresh
+// * Allow owner override
+
+contract EIP165Cache is Ownable, IEIP165Cache {
+
+
```
Comment (fabioberger) [inline packages/contracts/contracts/EIP165Cache.sol]: Can we remove this additional space?

File: packages/contracts/contracts/EIP165Cache.sol [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            let x := mload(0x40)   //Find empty storage location using "free memory pointer"
+            mstore(x, sig) //Place signature at begining of empty storage
+            mstore(add(x, 0x04), iface) //Place first argument directly next to signature
+
+            success := call(      //This is the critical change (Pop the top stack value)
+                30000,
+                addr, //To addr
+                0,    //No value
+                x,    //Inputs are stored at location x
+                0x8, //Inputs are 8 byes long
```
Comment (fabioberger) [inline packages/contracts/contracts/EIP165Cache.sol]: typo: `byes` -> `bytes`. Let's also simply have all comments separated from the code by a single space.

File: packages/contracts/contracts/EIP165Cache.sol [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            // TODO Could cache contract IEIP165 support
+            Status status = query(addr, interfaces[uint256(i)]);
+            if (status == Status.Supported) {
+                bitvector |= 1;
+            }
+        }
+        return bitvector;
+    }
+    
+    function refresh(address addr, bytes4 iface)
```
Comment (fabioberger) [inline packages/contracts/contracts/EIP165Cache.sol]: is `iface` short for `interface`? If so, can we simply use `interface` for readability?

**Comment by fabioberger** [comment]
This is awesome!

File: packages/contracts/contracts/EIP165Cache.sol (hunk 3 above)
Comment (Recmo) [reply packages/contracts/contracts/EIP165Cache.sol]: I wish. `interface` is a reserved keyword in Solidity. How about `signature` or `interfaceId`?

File: packages/contracts/contracts/EIP165Cache.sol (hunk 2 above)
Comment (Recmo) [reply packages/contracts/contracts/EIP165Cache.sol]: That's what I get for copy-pasting code :) I'll clean up this function a bit.

File: packages/contracts/contracts/EIP165Cache.sol (hunk 3 above)
Comment (willwarren89) [reply packages/contracts/contracts/EIP165Cache.sol]: +1 for `interfaceId`

**Pull request closed without merging**